bob-access/src/extractor.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{credentials::{Credentials, CredentialsBuilder}, error::Error};
+use axum::extract::RequestParts;
+use http::Request;
+use tonic::{transport::server::TcpConnectInfo, Request as TonicRequest};
+use super::CredentialsType;
+
+pub trait Extractor {
+    fn get_header(&self, header: &str) -> Result<Option<&str>, Error>;
+    fn get_extension<T: Send + Sync + 'static>(&self) -> Option<&T>;
+}
+
+pub trait ExtractorExt {
+    fn extract(&self, cred_type: CredentialsType) -> Result<Credentials, Error>;
+    fn extract_basic(&self) -> Result<Credentials, Error>;
+    fn extract_token(&self) -> Result<Credentials, Error>;
+    fn extract_internode(&self) -> Result<Credentials, Error>;
```

Comment:
This function is not needed as a part of trait

Reply:
To use this function inside extract_basic() and extract_token() it needs to be part of the trait (or should I create another trait, e.g. ExtractorInternode)

Reply:
Ok, then just integrate code from `extract_internode` directly into `extract_basic` and `extract_token`. Tiny code duplication in this situation is not a big deal